FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.JobManagement/AzureStorageJobQueueClient.cs

```diff
@@ -0,0 +1,717 @@
+                        jobLockEntities.First().PartitionKey,
+                        jobLockEntities.Select(entity => entity.RowKey).ToList()),
+                    cancellationToken: cancellationToken);
+
+                // step 4: get the existing jobInfo entities and job lock entities
+                jobLockEntities.Clear();
+                await foreach (var pageResult in jobLockEntityQueryResult.AsPages().WithCancellation(cancellationToken))
+                {
+                    jobLockEntities.AddRange(pageResult.Values);
+                }
+
+                if (jobLockEntities.Count != jobInfos.Count)
+                {
+                    _logger.LogError(ex, "Failed to add entities, some of the jobs already exist.");
+                    throw;
+                }
+
+                var jobInfoQueryResult = _azureJobInfoTableClient.QueryAsync<JobInfoEntity>(
+                    filter: TransactionGetByRowkeys(
+                        jobLockEntities.First().PartitionKey,
+                        jobLockEntities.Select(entity => entity.JobInfoEntityRowKey).ToList()),
+                    cancellationToken: cancellationToken);
+
+                jobInfoEntities.Clear();
+                await foreach (var pageResult in jobInfoQueryResult.AsPages().WithCancellation(cancellationToken))
+                {
+                    jobInfoEntities.AddRange(pageResult.Values);
+                }
+
+                _logger.LogInformation(ex, "Failed to add entities, the entities already exist. Fetched the existing jobs.");
+            }
+            catch (RequestFailedException ex) when (IsSpecifiedErrorCode(ex, AzureStorageErrorCode.RequestBodyTooLargeErrorCode))
+            {
+                _logger.LogError(ex, "The maximum size of a single table entity is 1MB, the size of definition is larger than 1MB.");
+                throw;
+            }
+
+            // step 5: try to add reverse index for jobInfo entity
+            try
+            {
+                transactionActions = jobInfoEntities.Select(jobInfoEntity =>
+                    new TableTransactionAction(TableTransactionActionType.Add, new JobReverseIndexEntity
+                    {
+                        PartitionKey = AzureStorageKeyProvider.JobReverseIndexPartitionKey(queueType, jobInfoEntity.Id),
+                        RowKey = AzureStorageKeyProvider.JobReverseIndexRowKey(queueType, jobInfoEntity.Id),
+                        JobInfoEntityPartitionKey = jobInfoEntity.PartitionKey,
+                        JobInfoEntityRowKey = jobInfoEntity.RowKey,
+                    }));
+
+                _ = await _azureJobInfoTableClient.SubmitTransactionAsync(transactionActions, cancellationToken);
+            }
+            catch (RequestFailedException ex) when (IsSpecifiedErrorCode(ex, AzureStorageErrorCode.AddEntityAlreadyExistsErrorCode))
+            {
+                _logger.LogInformation(ex, "The job reverse index entities already exist.");
```

Comment:
is that possible reverse index exist but queue message insert failed?

Reply:
Yes, will continue to insert the message in re-enqueue.

For the case, the message is enqueued and dequeued before updating the message info to table entity, we will skip the message if the message id in table entity is null at dequeue stage. unit test is added for it.